components/trezor_bridge/trezor_bridge_handler.cc

```diff
@@ -0,0 +1,184 @@
+    {"./data/messages/messages-v7.json",
+     IDR_TREZOR_BRIDGE_IFRAME_DATA_MESSAGES_MESSAGES_V7},
+    {"./data/messages/messages-v8.json",
+     IDR_TREZOR_BRIDGE_IFRAME_DATA_MESSAGES_MESSAGES_V8},
+    {"./data/firmware/1/releases.json",
+     IDR_TREZOR_BRIDGE_IFRAME_DATA_FIRMWARE_RELEASES1},
+    {"./data/firmware/2/releases.json",
+     IDR_TREZOR_BRIDGE_IFRAME_DATA_FIRMWARE_RELEASES2},
+    {"./data/bridge/releases.json",
+     IDR_TREZOR_BRIDGE_IFRAME_DATA_BRIDGE_RELEASES},
+};
+
+// TrezorSuite local address.
+const char kTrezorSuiteURL[] = "http://127.0.0.1:21325";
+
+net::NetworkTrafficAnnotationTag GetTrezorBridgeNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("trezor_bridge", R"(
+          semantics {
+            sender: "Trezor Bridge"
+            description:
+              "This bridge is used to communicate with TrezorSuite app "
+            trigger:
+              "Triggered by actions in brave://wallet."
+            data:
+              "Options of the commands."
+            destination: WEBSITE
+          }
+        )");
+}
+
+std::unique_ptr<network::SimpleURLLoader> CreateURLLoader(
+    const GURL& gurl,
+    const std::string& method,
+    const std::string& body) {
+  std::unique_ptr<network::ResourceRequest> request =
+      std::make_unique<network::ResourceRequest>();
+  request->url = gurl;
+  request->method = method;
+  // Use trezor.io origin because Trezor Sute accepts connections only from
+  // trezor.io domains.
+  request->headers.SetHeader(net::HttpRequestHeaders::kOrigin,
+                             "https://brave.trezor.io");
+  request->headers.SetHeader(net::HttpRequestHeaders::kContentType,
+                             "text/plain");
+
+  auto url_loader = network::SimpleURLLoader::Create(
+      std::move(request), GetTrezorBridgeNetworkTrafficAnnotationTag());
+  if (!body.empty())
+    url_loader->AttachStringForUpload(body, "text/plain");
+  return url_loader;
+}
+
+}  // namespace
+
+TrezorBridgeHandler::TrezorBridgeHandler(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory) {}
+
+TrezorBridgeHandler::~TrezorBridgeHandler() {}
+
+void TrezorBridgeHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "trezor-fetch",
+      base::BindRepeating(&TrezorBridgeHandler::HandleFetchRequest,
+                          base::Unretained(this)));
+}
+
+void TrezorBridgeHandler::HandleFetchRequest(const base::ListValue* args) {
+  std::string url;
+  CHECK(args->GetString(1, &url));
+  if (trezor_data_resources.contains(url)) {
+    auto resource_id = trezor_data_resources.at(url);
+    const ui::ResourceBundle& resource_bundle =
+        ui::ResourceBundle::GetSharedInstance();
+    const std::string& resource_text =
+        resource_bundle.LoadDataResourceString(resource_id);
+    RespondRequestCallback(args->GetList()[0], true, resource_text, "ok");
+    return;
+  }
```

Comment:
I think you can simplify and remove this part by adding those resources to the data source in the same way as the html files, and we can lean on regular webui data source to serve the large data. And then also filter out relative URLs from your `window.fetch` JS override.

If you create a GRD just for your component then iterate over the c++ generated map to add the source, you can control the relative url they will be served from, and you can provide a hint for yourself with the `resource_path` attribute, if neccessary:

```xml
<include name="IDR_TREZOR_BRIDGE_IFRAME_DATA_MESSAGES_MESSAGES_V6" file="${target_gen_dir}/messages/messages_v6.json" resource_path="messages/messages_v6.json" use_base_dir="false" type="BINDATA"/>
```

Reply:
doesnt work, the library uses window.fetch for json file

Reply:
Yes and I'm suggesting to clean up our workarounds by allowing it to fetch as normal for those resources, so instead of re-inventing the resource serving here, we simply add them to the data source, and if you make the resources in their own GRD file and grit target as suggested in my other comment, then you would have something like:

```cpp
  for (size_t i = 0; i < kTrezorBridgeStaticResourcesSize; ++i) {
    source->AddResourcePath(kTrezorBridgeStaticResources[i].resource_path, kTrezorBridgeStaticResources[i].id);
  }
```

Reply:
We just discovered this won't work because `window.fetch` does not support the `chrome` url scheme. Other workaround would be to use webpack to import the assets and avoid all the gn file copy and grit GRD etc, but the way you have it currently, serving the assets manually from the c++ message handler, is fine too IMO 👍

Reply:
doesnt work, discussed in slack